### PR TITLE
Various webgl fixes for framebuffer attachment test

### LINKED ISSUE #13710: webgl: uninitialized data leak from renderbufferStorage()
The FBO support I'm going to propose to merge won't do zero-initialization of renderbuffers initially.  This will certainly leak other graphics content from the process, but may also leak general user memory contents, depending on the driver implementation.  Since renderbuffers don't take in initial data like we do for textures, it's not easy to fix.  I came up with three possible solutions:
### Zero-initialize at creation using glClear().

This is invasive to the GL state: we need to make a new FBO, bind it, attach this renderbuffer to it, set the clear color/depth/stencil value, clear it, then restore the previous FBO state and clear color when we're done.
### Zero-initialize at first bind.

We could skip the FBO generation part of the previous solution if we delayed clearing until an FBO was bound with the renderbuffer attached.  All operations interacting with FBO contents (draws and readback) will check framebuffer status, and we go walking the attachments at that point, so we could just say "oh, an uninitialized RB at status check time?  Let's set the clear state, clear it, and restore clear state."
### Backing using textures instead

I had a thought that we could maybe just use GL textures to back WebGL renderbuffers, so we could zero-initialize in the same way as we do for WebGL textures.  However, unless we were to use packed depth/stencil, we couldn't represent stencil RBs as textures.

### LINKED ISSUE #20570: Audit that the framebuffer status is properly updated in WebGL
https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.6
https://www.khronos.org/registry/webgl/specs/latest/1.0/#6.7

### PR DESCRIPTION
**Pull request by jdm**
These changes resolve all panics on macOS when running framebuffer-object-attachment.html in headless and headful testing.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes OR
- [x] Fixes #13710. Fixes #20570.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21461)

### CODE REVIEW

**Comment by jdm** [comment]
Need to look at:
* /_mozilla/webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html
* /_mozilla/webgl/conformance-1.0.3/conformance/textures/texture-fakeblack.html

File: components/canvas/webgl_thread.rs [hunk 1]
```diff
@@ -18,6 +18,33 @@ use webrender_api;
...
+}
+
+pub struct GLState {
+    clear_color: (f32, f32, f32, f32),
+    scissor_test_enabled: bool,
+    stencil_write_mask: (u32, u32),
+    stencil_clear_value: i32,
+    depth_write_mask: bool,
+    depth_clear_value: f64,
+}
```
Comment (nox) [inline components/canvas/webgl_thread.rs]: Why is this on the GL side when it could be on the DOM side?

Comment (nox) [reply components/canvas/webgl_thread.rs]: I'm strongly against caching *some* stuff on the GL side and *some* stuff on the DOM side. All of the bookkeeping I've implemented this year has been put on the DOM side to be able, I see no reason to put those values on the GL side.

Comment (nox) [reply components/canvas/webgl_thread.rs]: Never mind I see why now.

File: components/canvas/webgl_thread.rs [hunk 2]
```diff
@@ -1091,6 +1172,63 @@ impl WebGLImpl {
         assert_eq!(error, gl::NO_ERROR, "Unexpected WebGL error: 0x{:x} ({})", error, error);
     }
 
+    #[allow(unsafe_code)]
```
Comment (nox) [inline components/canvas/webgl_thread.rs]: This doesn't use any unsafe code.

File: components/script/dom/webglrenderingcontext.rs [hunk 3]
```diff
@@ -200,6 +200,7 @@ pub struct WebGLRenderingContext {
     current_scissor: Cell<(i32, i32, i32, i32)>,
     #[ignore_malloc_size_of = "Because it's small"]
     current_clear_color: Cell<(f32, f32, f32, f32)>,
+    #[ignore_malloc_size_of = "Because it's small"]
```
Comment (nox) [inline components/script/dom/webglrenderingcontext.rs]: Why is this suddenly needed?

**Comment by jdm** [comment]
Note to self - DeleteRenderbuffer already contains some code to deal with deleting a renderbuffer that is bound to the current framebuffer. Need to figure out how to merge that with my reimplementation in this PR.

**Comment by jdm** [comment]
This is ready for review.

File: components/script/dom/webglrenderbuffer.rs [hunk 4]
```diff
@@ -24,6 +26,8 @@ pub struct WebGLRenderbuffer {
     size: Cell<Option<(i32, i32)>>,
     internal_format: Cell<Option<u32>>,
     is_initialized: Cell<bool>,
+    // Framebuffer that this texture is attached to.
+    attached_framebuffers: DomRefCell<Vec<Dom<WebGLFramebuffer>>>,
```
Comment (nox) [inline components/script/dom/webglrenderbuffer.rs]: I don't understand why this is needed and why we can't just keep track of a refcount.

File: tests/wpt/webgl/meta/conformance2/rendering/framebuffer-texture-level1.html.ini [hunk 5]
```diff
@@ -8,3 +8,9 @@
   [WebGL test #2: gl.checkFramebufferStatus(gl.FRAMEBUFFER) should be 36053. Was 36061.]
     expected: FAIL
 
+  [WebGL test #2: gl.checkFramebufferStatus(gl.FRAMEBUFFER) should be 36053. Was 36055.]
+    expected: FAIL
+
+  [WebGL test #1: gl.checkFramebufferStatus(gl.FRAMEBUFFER) should be 36054. Was 36055.]
+    expected: FAIL
+
```
Comment (nox) [inline tests/wpt/webgl/meta/conformance2/rendering/framebuffer-texture-level1.html.ini]: The previous failures are now outdated and should be removed.

**Review by nox — CHANGES_REQUESTED** [review]
An expectation nit, and a question about some binding tracking.

**Comment by jdm** [comment]
You were correct; it wasn't necessary to track the attached framebuffers. The resulting code is easier to follow, which is nice.

**Comment by jdm** [comment]
4edb7b1 is the only commit that's new.

**Comment by nox** [comment]
The following expectations reference #13710:

* `tests/wpt/webgl/meta/conformance/misc/uninitialized-test.html.ini`
* `tests/wpt/webgl/meta/conformance/renderbuffers/depth-renderbuffer-initialization.html.ini`
* `tests/wpt/webgl/meta/conformance/renderbuffers/renderbuffer-initialization.html.ini`
* `tests/wpt/webgl/meta/conformance/renderbuffers/stencil-renderbuffer-initialization.html.ini`

AFAICT they are not all changed in your PR. Could you reenable them? Looks good to me otherwise, r=me.

**Comment by nox** [comment]
@jdm Could you disable the tests about uninitialised buffers again and make this land?

**Comment by jdm** [comment]
@bors-servo retry
* #21493 21493

**Pull request merged**